### PR TITLE
update rendering details

### PR DESCRIPTION
**Pull request by vmpstr**

### CODE REVIEW

**Comment by vmpstr** [comment]
Please take a look.

File: explainers/update-rendering.md [hunk 1]
```diff
@@ -42,12 +48,49 @@ are ready to be displayed to screen.  Note that repeated calls to
 updateRendering() returns new promises, with the ultimate priority of the
 request being the highest requested priority.
 
+### Proposal: `keepUpdated` element attribute
+The declarative version `updateRendering()` is an attribute called `keepUpdated`.
```
Comment (chrishtr) [inline explainers/update-rendering.md:52]:
```suggestion
The declarative version of `updateRendering()` is an attribute called `keepUpdated`.
```

File: explainers/update-rendering.md [hunk 2]
```diff
@@ -16,9 +19,12 @@ words, if the website intends to show an element whose contents are currently
 skipped, then skipping work may cause jank when the contents are ultimately
 presented.
 
-### Proposal
-The proposal is to add the following function, which lets the user-agent know
-that the element’s contents should be updated in preparation for display:
+### Proposal: `updateRendering()` JavaScript function
```
Comment (chrishtr) [inline explainers/update-rendering.md:22]: WDYT of just removing updateRendering() for now, and just say "an imperative equivalent to keepUpdated could be added in the future".

Comment (chrishtr) [reply explainers/update-rendering.md:22]: I am pretty sure the attribute will be an easier sell for standards, just as it was for FB.

Comment (vmpstr) [reply explainers/update-rendering.md:22]: Sounds good.

File: explainers/update-rendering.md [hunk 3]
```diff
@@ -1,13 +1,16 @@
-## `updateRendering()`
+## `updateRendering()` and `keepUpdated`
 
 ### TL;DR
 updateRendering() is an asynchronous JavaScript function on Element which
 requests that the rendering for the element be updated in preparation for
 display.
 
+The declarative counterpart is a `keepUpdated` attribute which lets the
```
[the remark this answers is not in the document]
Comment (vmpstr) [reply explainers/update-rendering.md:8]: Done.

**Comment by vmpstr** [comment]
PTAL

File: explainers/update-rendering.md [hunk 4]
```diff
@@ -16,38 +22,62 @@ words, if the website intends to show an element whose contents are currently
...
+  does _not_ include elements that themselves have `content-visibility: hidden`
+  style, since their rendering is updated and the attribute would cause the
+  subtree to be udpated as well.
 
+* Setting the attribute on an element that contains descendants with
+  `content-visibility: hidden`, `display: none`, or similar styles would not
+  cause such descendants to be updated. This is a consequence of the fact that
+  a fully updated parent element has all its rendering work completed without
+  updating such descendants. ([more
+  details](https://github.com/WICG/display-locking/issues/196))
```
Comment (chrishtr) [inline explainers/update-rendering.md:82]:
```suggestion
  details](https://github.com/WICG/display-locking/issues/196)). If such an update is desired, a `renderPriority` attribute should be set on such elements.
```

Should we add a sentence saying something like "a recursive `renderPriority` setting may be considered in the future?

File: explainers/update-rendering.md [hunk 5]
```diff
@@ -1,13 +1,19 @@
-## `updateRendering()`
+## `renderPriority` attribute.
+
+_Note that issues referenced in this document can refer to either the
+`renderPriority` attribute or `updateRendering()` JavaScript API. The issues
```
Comment (chrishtr) [inline explainers/update-rendering.md:4]: is updateRendering mentioned anywhere in this doc now? If not, remove?

File: explainers/update-rendering.md [hunk 6]
```diff
@@ -16,38 +22,62 @@ words, if the website intends to show an element whose contents are currently
...
+  for the purposes of User Agent prioritization.
+
+Note that the attribute values follow [the postTask API](https://wicg.github.io/scheduling-apis/#sec-task-priorities)
+example.
+
+
+### Interaction with `content-visibility`
+Since the User Agent typically keeps rendering state of subtrees up-to-date,
+this feature would be a no-op in a majority of cases. For instance, having
+`renderPriority` on a visible, on screen, element would not have to do any
```
[the remark this answers is not in the document]
Comment (vmpstr) [reply explainers/update-rendering.md:53]: I'm not sure about that.. I guess we haven't really discussed what userVisible means, but even if say it changes style and layout in separate frames, then that seems like it may be web breaking. I'm worried about making this enter the "render async" territory where things are visible but are not updated synchronously.

Or do you think that this is the opportunity to change this?

File: explainers/update-rendering.md (hunk 5 above)
Comment (vmpstr) [reply explainers/update-rendering.md:4]: It's mentioned in the issues which I reference, so I just wanna keep this here as a clarification that we're talking about one and the same

**Comment by vmpstr** [comment]
PTAL

File: explainers/update-rendering.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (chrishtr) [reply explainers/update-rendering.md:58]: > Or do you think that this is the opportunity to change this?

I think it's an opportunity to consider introducing async rendering across the board, not just in content-visibility. It would not be render breaking for existing content, because the `auto` behavior would be unchanged from today. We would also give UAs wiggle room to not implement it (e.g. we'd likely not do so in Chromium at first).

File: explainers/update-rendering.md (hunk 5 above)
Comment (chrishtr) [reply explainers/update-rendering.md:4]: ok

File: explainers/update-rendering.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (vmpstr) [reply explainers/update-rendering.md:58]: I suggest that we move some of this discussion into an issue. I think that if a developer uses this in conjunction with `content-visibility`, and then at some point they remove `content-visibility`, they might not expect that although the element is now visible, it tends to be updated asynchronously:

```js
<div id=e style="content-visibility: hidden" renderPriority=background> ... </div>
...
e.style = "";
// e is now maybe asynchronous?
```

I think this text can change for sure, but coupling these two features in this explainer might complicated TAG/standards reviews. WDYT?

[the remark this answers is not in the document]
Comment (chrishtr) [reply explainers/update-rendering.md:58]: ok sure let's put it in an issue. File one and link to it, with a reference to the behavior being an open discussion point?

[the remark this answers is not in the document]
Comment (vmpstr) [reply explainers/update-rendering.md:58]: SGTM

[the remark this answers is not in the document]
Comment (vmpstr) [reply explainers/update-rendering.md:58]: Done.

**Review by chrishtr — APPROVED** [review]
